File: BackEnd/DotNet/WebDigital/WebDigital/Services/Implements/CartService.cs
using Microsoft.EntityFrameworkCore;
using WebDigital.dto;
using WebDigital.Exceptions;
using WebDigital.Models;
using WebDigital.Services.Interface;

namespace WebDigital.Services.Implements
{
    public class CartService : ICartService
    {
        private readonly BookwormfinalBackupContext _context;

        public CartService(BookwormfinalBackupContext context)
        {
            _context = context;
        }

        public async Task<List<CartItemDto>> GetCart(int userId)
        {
            return await _context.Carts
                .Include(c => c.Product)
                .Where(c => c.UserId == userId)
                .Select(c => new CartItemDto
                {
                    CartId = c.CartId,
                    ProductId = c.ProductId,
                    ProductName = c.Product.ProductName,
                    ProductImage = c.Product.ProductImage,
                    ProductBaseprice = c.Product.ProductBaseprice,
                    ProductOfferprice = c.Product.ProductOfferprice,
                    Qty = c.Qty,
                    LineTotal = (c.Product.ProductOfferprice ?? c.Product.ProductBaseprice) * c.Qty,
                    IsRentable = c.Product.IsRentable == 1
                })
                .ToListAsync();
        }

        public async Task<int> AddToCart(int userId, CartAddDto dto)
        {
            var productExists = await _context.Products.AnyAsync(p => p.ProductId == dto.ProductId);
            if (!productExists)
                throw new NotFoundException($"Product with id {dto.ProductId} was not found.");

            if (dto.Qty < 1)
                throw new BadRequestException("Quantity must be at least 1.");

            // If this product is already in the user's cart, just increase the quantity
            var existing = await _context.Carts
                .FirstOrDefaultAsync(c => c.UserId == userId && c.ProductId == dto.ProductId);

            if (existing != null)
            {
                existing.Qty += dto.Qty;
                await _context.SaveChangesAsync();
                return existing.CartId;
            }

            var cartItem = new Cart
            {
                UserId = userId,
                ProductId = dto.ProductId,
                Qty = dto.Qty
            };

            _context.Carts.Add(cartItem);
            await _context.SaveChangesAsync();

            return cartItem.CartId;
        }

        public async Task UpdateQuantity(int userId, int cartId, CartUpdateDto dto)
        {
            if (dto.Qty < 1)
                throw new BadRequestException("Quantity must be at least 1.");

            var cartItem = await _context.Carts
                .FirstOrDefaultAsync(c => c.CartId == cartId && c.UserId == userId);

            if (cartItem == null)
                throw new NotFoundException("Cart item not found.");

            cartItem.Qty = dto.Qty;
            await _context.SaveChangesAsync();
        }

        public async Task RemoveItem(int userId, int cartId)
        {
            var cartItem = await _context.Carts
                .FirstOrDefaultAsync(c => c.CartId == cartId && c.UserId == userId);

            if (cartItem == null)
                throw new NotFoundException("Cart item not found.");

            _context.Carts.Remove(cartItem);
            await _context.SaveChangesAsync();
        }

        public async Task ClearCart(int userId)
        {
            var items = await _context.Carts.Where(c => c.UserId == userId).ToListAsync();
            _context.Carts.RemoveRange(items);
            await _context.SaveChangesAsync();
        }
    }
}

File: BackEnd/Java/transaction-service/src/main/java/com/digitalshelf/transactionservice/service/CartService.java
package com.digitalshelf.transactionservice.service;

import com.digitalshelf.transactionservice.client.ProductClient;
import com.digitalshelf.transactionservice.dto.CartAddDto;
import com.digitalshelf.transactionservice.dto.CartItemDto;
import com.digitalshelf.transactionservice.dto.CartUpdateDto;
import com.digitalshelf.transactionservice.dto.ProductSummaryDto;
import com.digitalshelf.transactionservice.entity.Cart;
import com.digitalshelf.transactionservice.exception.BadRequestException;
import com.digitalshelf.transactionservice.exception.NotFoundException;
import com.digitalshelf.transactionservice.repository.CartRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;

@Service
@RequiredArgsConstructor
public class CartService {

    private final CartRepository cartRepository;
    private final ProductClient productClient;

    public List<CartItemDto> getCart(Integer userId) {
        return cartRepository.findByUserId(userId).stream()
                .map(cart -> {
                    ProductSummaryDto product = productClient.getProductSummary(cart.getProductId());
                    BigDecimal unitPrice = product.getProductOfferprice() != null
                            ? product.getProductOfferprice() : product.getProductBaseprice();

                    return CartItemDto.builder()
                            .cartId(cart.getCartId())
                            .productId(cart.getProductId())
                            .productName(product.getProductName())
                            .productImage(product.getProductImage())
                            .productBaseprice(product.getProductBaseprice())
                            .productOfferprice(product.getProductOfferprice())
                            .qty(cart.getQty())
                            .lineTotal(unitPrice.multiply(BigDecimal.valueOf(cart.getQty())))
                            .isRentable(product.isRentable())
                            .build();
                })
                .toList();
    }

    public Integer addToCart(Integer userId, CartAddDto dto) {
        // Confirms the product genuinely exists (Feign call throws if product-service 404s)
        productClient.getProductSummary(dto.getProductId());

        if (dto.getQty() < 1) {
            throw new BadRequestException("Quantity must be at least 1.");
        }

        var existing = cartRepository.findByUserIdAndProductId(userId, dto.getProductId());

        if (existing.isPresent()) {
            Cart cart = existing.get();
            cart.setQty(cart.getQty() + dto.getQty());
            cartRepository.save(cart);
            return cart.getCartId();
        }

        Cart cart = new Cart();
        cart.setUserId(userId);
        cart.setProductId(dto.getProductId());
        cart.setQty(dto.getQty());
        cartRepository.save(cart);
        return cart.getCartId();
    }

    public void updateQuantity(Integer userId, Integer cartId, CartUpdateDto dto) {
        if (dto.getQty() < 1) {
            throw new BadRequestException("Quantity must be at least 1.");
        }

        Cart cart = cartRepository.findByCartIdAndUserId(cartId, userId)
                .orElseThrow(() -> new NotFoundException("Cart item not found."));

        cart.setQty(dto.getQty());
        cartRepository.save(cart);
    }

    public void removeItem(Integer userId, Integer cartId) {
        Cart cart = cartRepository.findByCartIdAndUserId(cartId, userId)
                .orElseThrow(() -> new NotFoundException("Cart item not found."));
        cartRepository.delete(cart);
    }

    public void clearCart(Integer userId) {
        cartRepository.deleteAll(cartRepository.findByUserId(userId));
    }
}